Hot reload fails when using BeanParam

### Describe the bug

Every usage of BeanParam causes hot reload to fail and requires manual reload of the app.
I'm running graalvm 23

### Expected behavior

When using BeanParam for the first time, everything works as expected including after a small change in the code when the hot reload kicks in. 

### Actual behavior

When I change anything in the code while the app is running, the hot reload fails on the next request to the service with the following (shortened) exception:
```
loader constraint violation: when resolving method 'com.freifeld.SimpleBeanParam com.freifeld.SimpleBeanParam.__quarkus_rest_inject(org.jboss.resteasy.reactive.server.injection.ResteasyReactiveInjectionContext)' the class loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @52af26ee of the current class, org/jboss/resteasy/reactive/server/core/parameters/RecordBeanParamExtractor, and the class loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @e30e4ad for the method's defining class, com/freifeld/SimpleBeanParam, have different Class objects for the type com/freifeld/SimpleBeanParam used in the signature (org.jboss.resteasy.reactive.server.core.parameters.RecordBeanParamExtractor is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @52af26ee, parent loader 'app'; com.freifeld.SimpleBeanParam is in unnamed module of loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @e30e4ad, parent loader io.quarkus.bootstrap.classloading.QuarkusClassLoader @52af26ee)
```

To make it work again, it requires a manual restart.


### How to Reproduce?

https://github.com/FreifeldRoyi/bean-param-issue

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

3.17.6

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

I am experiencing this same issue with a record class annotated with @BeanParam in the @POST REST method.  Initial startup with "quarkus dev" allows the REST call to succeed when called multiple times, but a code change triggering dynamic reload of running system results in an invocation target exception.  Each field specified in the record parameters is annotated with @RestForm.

Some relevant parts of stack trace:

java.lang.RuntimeException: Failed to find target generated factory method on record @BeanParam type
	at org.jboss.resteasy.reactive.server.core.parameters.RecordBeanParamExtractor.<init>(RecordBeanParamExtractor.java:19)
....
Caused by: java.lang.IllegalAccessException: no such method: com.lifelenstech.bundlesubmission.rest.BundleForm.__quarkus_rest_inject(ResteasyReactiveInjectionContext)BundleForm/invokeStatic

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/beanparam/BeanParamRecordDevModeTest.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/beanparam/BeanParamRecordDevModeTest.java
@@ -1,0 +1,42 @@
+package io.quarkus.resteasy.reactive.server.test.beanparam;
+
+import static io.restassured.RestAssured.when;
+import static org.hamcrest.CoreMatchers.is;
+
+import java.util.function.Supplier;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusDevModeTest;
+
+public class BeanParamRecordDevModeTest {
+
+    @RegisterExtension
+    static QuarkusDevModeTest TEST = new QuarkusDevModeTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class).addClasses(FirstAndSecondResource.class,
+                            FirstAndSecondResource.Param.class);
+                }
+            });
+
+    @Test
+    void test() {
+        when().get("fs/foo/bar")
+                .then()
+                .statusCode(200)
+                .body(is("foo-bar"));
+
+        TEST.modifySourceFile(FirstAndSecondResource.class, (orig) -> orig.replace("-", "#"));
+
+        when().get("fs/foo/bar")
+                .then()
+                .statusCode(200)
+                .body(is("foo#bar"));
+    }
+
+}

--- a/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/beanparam/FirstAndSecondResource.java
+++ b/extensions/resteasy-reactive/rest/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/beanparam/FirstAndSecondResource.java
@@ -1,0 +1,20 @@
+package io.quarkus.resteasy.reactive.server.test.beanparam;
+
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import org.jboss.resteasy.reactive.RestPath;
+
+@Path("fs")
+public class FirstAndSecondResource {
+
+    @Path("{first}/{second}")
+    @GET
+    public String firstAndSecond(Param param) {
+        return param.first() + "-" + param.second();
+    }
+
+    public record Param(@RestPath String first, @RestPath String second) {
+
+    }
+}

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/RecordBeanParamExtractor.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/RecordBeanParamExtractor.java
@@ -9,7 +9,7 @@ import org.jboss.resteasy.reactive.server.injection.ResteasyReactiveInjectionCon
 
 public class RecordBeanParamExtractor implements ParameterExtractor {
 
-    private MethodHandle factoryMethod;
+    private final MethodHandle factoryMethod;
 
     public RecordBeanParamExtractor(Class<?> target) {
         try {

--- a/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/RecordBeanParamExtractor.java
+++ b/independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/core/parameters/RecordBeanParamExtractor.java
@@ -2,7 +2,6 @@ package org.jboss.resteasy.reactive.server.core.parameters;
 
 import java.lang.invoke.MethodHandle;
 import java.lang.invoke.MethodHandles;
-import java.lang.invoke.MethodType;
 
 import org.jboss.resteasy.reactive.server.core.ResteasyReactiveRequestContext;
 import org.jboss.resteasy.reactive.server.injection.ResteasyReactiveInjectionContext;
@@ -13,8 +12,8 @@ public class RecordBeanParamExtractor implements ParameterExtractor {
 
     public RecordBeanParamExtractor(Class<?> target) {
         try {
-            factoryMethod = MethodHandles.lookup().findStatic(target, "__quarkus_rest_inject",
-                    MethodType.methodType(target, ResteasyReactiveInjectionContext.class));
+            factoryMethod = MethodHandles.lookup()
+                    .unreflect(target.getMethod("__quarkus_rest_inject", ResteasyReactiveInjectionContext.class));
         } catch (NoSuchMethodException | IllegalAccessException e) {
             throw new RuntimeException("Failed to find target generated factory method on record @BeanParam type", e);
         }
@@ -27,7 +26,6 @@ public class RecordBeanParamExtractor implements ParameterExtractor {
         } catch (RuntimeException e) {
             throw e;
         } catch (Throwable e) {
-            e.printStackTrace();
             throw new RuntimeException("Failed to invoke generated factory method on record @BeanParam type", e);
         }
     }

